desc/objectives/_stability.py

```diff
@@ -367,3 +369,211 @@ def compute(self, params, constants=None):
             profiles=constants["profiles"],
         )
         return constants["transforms"]["grid"].compress(data["magnetic well"])
+
+
+class BallooningStability(_Objective):
+    """A type of ideal MHD instability.
+
+    Infinite-n ideal MHD ballooning modes are of significant interest.
+    These instabilities are also related to smaller-scale kinetic instabilities.
+    With this class, we optimize MHD equilibria against the ideal ballooning mode.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    target : {float, ndarray, callable}, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        Must be broadcastable to Objective.dim_f. If a callable, should take a
+        single argument `rho` and return the desired value of the profile at those
+        locations.
+    bounds : tuple of {float, ndarray, callable}, optional
+        Lower and upper bounds on the objective. Overrides target.
+        Both bounds must be broadcastable to to Objective.dim_f
+        If a callable, each should take a single argument `rho` and return the
+        desired bound (lower or upper) of the profile at those locations.
+    weight : {float, ndarray}, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        Must be broadcastable to to Objective.dim_f
+    normalize : bool, optional
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool, optional
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+    grid : Grid, optional
+        Collocation grid containing the nodes to evaluate at.
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    _coordinates = "r"
+    _units = "(normalized)"
+    _print_value_fmt = "Ideal-ballooning Stability: {:10.3e} "
+
+    def __init__(
+        self,
+        eq=None,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        deriv_mode="rev",
+        loss_function=None,
+        rho=0.5,
+        alpha=jnp.array([[0.0], [0.2], [0.4], [0.6], [0.8], [1.0], [1.2], [1.4]]),
+        zetamax=3 * jnp.pi,
+        nzeta=200,
```

Comment:
these args aren't in the docstring above. Also, with the changes in #1024 can the resolution stuff be handled by a grid in `raz` coordinates?

Reply:
I'd prefer not to make this part of the public API if we know we'll be changing it soon. At the very least you could make it take a grid in `raz` as input and then just get the rho, alpha, zeta values from that rather than having them be specified manually?

Also in the compute functions it looks like its hard-coded for 8 alphas, that should definitely be generalized if you're allowing the user to specify the alpha values

Reply:
I'd recommend copying the effective ripple or Gamma_c objective and changing `Gamma_c` to `ideal_ball_gamma2` in the keys to compute.

Reply:
> I'd prefer not to make this part of the public API if we know we'll be changing it soon. At the very least you could make it take a grid in raz as input and then just get the rho, alpha, zeta values from that rather than having them be specified manually?

The user can specify an alpha and it will be expanded or shrunk to N_alpha = 8. For rho, I can add an assert statement.
The values above are just defaults. User can specify whatever they want.

> I'd recommend copying the effective ripple or Gamma_c objective and changing Gamma_c to ideal_ball_gamma2 in the keys to compute.

It's not as simple as that. A lot of the tests fail because of things like map_coordinates, source grid in compute functions, and how you changed test_compute_everything. Also, the APIs are fairly different.

Reply:
> map_coordinates
I did have to change map_coordinates in the ripple branch to fix some stuff with map_coordinates using jax arrays. In that case I assume you've run into similar errors, so sure we can change later after this PR.

> test_compute_everything

If you have issues I can help debug in this [thread](https://github.com/PlasmaControl/DESC/pull/1170#discussion_r1717285910)